lib/evmone/vm.hpp

```diff
@@ -51,22 +53,18 @@ class VMTracer
     virtual void onEndExecution() noexcept = 0;
 };
 
-/// The evmone EVMC instance.
-class VM : public evmc_vm
+void TracerListNode::add_tracer(std::unique_ptr<VMTracer> tracer) noexcept
 {
-    std::unique_ptr<VMTracer> m_tracer;
+    if (m_next_tracer)
+        m_next_tracer->add_tracer(std::move(tracer));
+    else
+        m_next_tracer = std::move(tracer);
+}
 
+/// The evmone EVMC instance.
+class VM : public evmc_vm, public TracerListNode
```

Comment:
I would better inherit privately from `TraceListNode` or even make it a member, and add delegating public `add_tracer` method.
Otherwise it can be incorrectly deleted via `TracerListNode*` which doesn't have virtual destructor.

But maybe the same is true about deleting via `evmc_vm*`, so don't know how much it is of a concern.

Reply:
This is not a concern, because `VM` is not polymorphic type. But I will try this, because it will give me better method name.